File: js/extract.js
// NeoRender — Auto-extraction helpers
// Injected into V8 runtime. Called from Rust extract module.

globalThis.__neo_extract_tables = function() {
    const tables = [];
    for (const table of document.querySelectorAll('table')) {
        const headers = [...table.querySelectorAll('th')].map(th => th.textContent?.trim() || '');
        const rows = [...table.querySelectorAll('tr')].map(tr =>
            [...tr.querySelectorAll('td')].map(td => td.textContent?.trim() || '')
        ).filter(r => r.length > 0);
        if (rows.length > 0) tables.push({headers, rows});
    }
    return JSON.stringify(tables);
};

globalThis.__neo_extract_article = function() {
    // Find the main article content
    const article = document.querySelector('article')
        || document.querySelector('[role="main"]')
        || document.querySelector('main')
        || document.body;
    const title = document.querySelector('h1')?.textContent?.trim() || document.title;

    // Extract clean text (skip nav, aside, footer)
    const body = [];
    function walk(el) {
        if (!el || !el.tagName) return;
        const tag = el.tagName.toLowerCase();
        if (['script','style','nav','aside','footer','header','noscript','svg'].includes(tag)) return;
        if (el.getAttribute?.('role') === 'navigation') return;
        for (const child of el.childNodes || []) {
            if (child.nodeType === 3) {
                const t = child.textContent?.trim();
                if (t) body.push(t);
            } else walk(child);
        }
        if (['p','h1','h2','h3','h4','h5','h6','li','br'].includes(tag)) body.push('\n');
    }
    walk(article);

    // Extract meta
    const author = document.querySelector('[name="author"]')?.content
        || document.querySelector('[rel="author"]')?.textContent?.trim()
        || '';
    const date = document.querySelector('time')?.getAttribute('datetime')
        || document.querySelector('[name="date"]')?.content
        || '';

    return JSON.stringify({
        title,
        author,
        date,
        body: body.join(' ').replace(/\n{3,}/g, '\n\n').trim()
    });
};

globalThis.__neo_extract_form_schema = function(selector) {
    const form = selector ? document.querySelector(selector) : document.querySelector('form');
    if (!form) return JSON.stringify(null);

    const fields = [];
    for (const el of form.querySelectorAll('input, select, textarea')) {
        const field = {
            name: el.name || el.id || '',
            type: el.type || el.tagName.toLowerCase(),
            required: el.required || el.hasAttribute?.('required'),
            placeholder: el.placeholder || '',
            value: el.value || '',
        };
        if (el.tagName === 'SELECT') {
            field.options = [...el.querySelectorAll('option')].map(o => ({
                value: o.value,
                text: o.textContent?.trim()
            }));
        }
        if (field.name) fields.push(field);
    }

    return JSON.stringify({
        action: form.action || '',
        method: (form.method || 'GET').toUpperCase(),
        fields
    });
};

globalThis.__neo_extract_structured = function() {
    // JSON-LD
    const jsonld = [];
    for (const script of document.querySelectorAll('script[type="application/ld+json"]')) {
        try { jsonld.push(JSON.parse(script.textContent)); } catch {}
    }
    // Open Graph
    const og = {};
    for (const meta of document.querySelectorAll('meta[property^="og:"]')) {
        og[meta.getAttribute('property')] = meta.content;
    }
    return JSON.stringify({jsonld, og});
};

File: src/neorender/extract.rs
//! Auto-extraction — tables, articles, form schemas, structured data.
//!
//! Calls JS helpers injected via js/extract.js and returns parsed JSON.

use deno_core::JsRuntime;
use serde_json::Value;

/// Extract all HTML tables as structured JSON.
/// Returns: `[{headers: [string], rows: [[string]]}]`
pub fn extract_tables(runtime: &mut JsRuntime) -> Result<Vec<Value>, String> {
    let json_str = eval_string(runtime, "__neo_extract_tables()")?;
    serde_json::from_str(&json_str)
        .map_err(|e| format!("extract_tables parse error: {e}"))
}

/// Extract article content: title, author, date, body text.
/// Returns: `{title, author, date, body}`
pub fn extract_article(runtime: &mut JsRuntime) -> Result<Value, String> {
    let json_str = eval_string(runtime, "__neo_extract_article()")?;
    serde_json::from_str(&json_str)
        .map_err(|e| format!("extract_article parse error: {e}"))
}

/// Extract form schema for the first (or specified) form.
/// Returns: `{action, method, fields: [{name, type, required, placeholder, value, options?}]}`
/// Returns `null` Value if no form found.
pub fn extract_form_schema(runtime: &mut JsRuntime, selector: Option<&str>) -> Result<Value, String> {
    let js = match selector {
        Some(sel) => {
            let escaped = sel.replace('\'', "\\'").replace('\\', "\\\\");
            format!("__neo_extract_form_schema('{}')", escaped)
        }
        None => "__neo_extract_form_schema()".to_string(),
    };
    let json_str = eval_string(runtime, &js)?;
    serde_json::from_str(&json_str)
        .map_err(|e| format!("extract_form_schema parse error: {e}"))
}

/// Extract structured data: JSON-LD and Open Graph meta.
/// Returns: `{jsonld: [object], og: {property: value}}`
pub fn extract_structured(runtime: &mut JsRuntime) -> Result<Value, String> {
    let json_str = eval_string(runtime, "__neo_extract_structured()")?;
    serde_json::from_str(&json_str)
        .map_err(|e| format!("extract_structured parse error: {e}"))
}

// ─── Helpers ───

fn eval_string(runtime: &mut JsRuntime, js: &str) -> Result<String, String> {
    let result = runtime.execute_script("<neo:extract>", js.to_string())
        .map_err(|e| format!("extract eval error: {e}"))?;
    let scope = &mut runtime.handle_scope();
    let local = deno_core::v8::Local::new(scope, result);
    if let Some(s) = local.to_string(scope) {
        Ok(s.to_rust_string_lossy(scope))
    } else {
        Ok("null".to_string())
    }
}
